### PR TITLE
Resetting current directory from unwritable "/" to user home in CoqIDE MacOS released package

### PR DESCRIPTION
**Pull request by herbelin**
**Kind:** Enhancement / fix

At startup, the MacOS CoqIDE package has its current working directory (cwd) set to `/`, resulting in write errors when using commands that work by default on the cwd. Such commands include `Extraction "foo.ml" id`, `Print Universes "file"`, `Redirect` (see e.g. [#4953](https://github.com/coq/coq/issues/4953#issuecomment-337545391) for examples).

This PR addresses the issue by changing it "conventionally" to the home directory. It is not necessarily what we want but it is writable and this avoids surprises for newcomers, see e.g.:
- https://coq.discourse.group/t/extraction-system-error/1399/5
- https://stackoverflow.com/questions/34349290/coq-extraction-permission-denied

Probably as important would be that commands that write to the current directory, like extraction, or `Redirect`, or `Print Universes` tell where they write and that this is configurable (see e.g. #4953, #9148, #8649).

Note that vscoq on MacOS has the same issue (@maximedenes, @fakusb).

- [ ] Added **changelog**.

### CODE REVIEW

**Comment by herbelin** [comment]
@coq/nix-maintainers: on Nix, getting the home directory seems to fail (using `Envars.home`). Any idea of what default path would be relevant for Coq user interfacses. Such a default path is relevant for instance for the `Extraction "foo.ml" id` command.

**Comment by SkySkimmer** [comment]
>@coq/nix-maintainers: on Nix, getting the home directory seems to fail (using Envars.home). Any idea of what default path would be relevant for Coq user interfacses. Such a default path is relevant for instance for the Extraction "foo.ml" id command.

`Filename.get_temp_dir_name`

File: vernac/vernacentries.ml [hunk 1]
```diff
@@ -1656,6 +1651,15 @@ let () =
...
       optwrite = CDebug.set_flags }
 
+let pwd_key = ["Pwd"]
+
+let () =
+  declare_string_option
+    { optdepr  = false;
+      optkey   = pwd_key;
+      optread  = Sys.getcwd;
+      optwrite = set_pwd }
```
Comment (SkySkimmer) [inline vernac/vernacentries.ml:1661]: needs doc
Also not sure if it makes sense to have this be an option, do we really want the possibility of Require doing chdir?

Comment (herbelin) [reply vernac/vernacentries.ml:1661]: > do we really want the possibility of Require doing chdir?

I would say no.

But there is still the question of synchronization with backtracking. E.g. if I do:
```coq
Goal True.
Pwd.
Cd "/".
Pwd.
```
Then going to the end then backtracking on the first `Pwd` gives `/`.

> needs doc

Yes, this should be part of #14720.

**Comment by herbelin** [comment]
> > @coq/nix-maintainers: on Nix, getting the home directory seems to fail (using Envars.home). Any idea of what default path would be relevant for Coq user interfacses. Such a default path is relevant for instance for the Extraction "foo.ml" id command.
> 
> `Filename.get_temp_dir_name`

@SkySkimmer : actually, I don't know what cause the NIX failure. Is there a way to get the output of the failing `ide` tests on Nix? Or to rerun them with `PRINT_LOGS=1`?

**Comment by SkySkimmer** [comment]
It's in the artifacts.

**Comment by herbelin** [comment]
> It's in the artifacts.

Thanks, and sorry for having forgot this information once more. CI should now work on Nix.

**Comment by vbgl** [comment]
If I may share my thoughts, the issues that are referred to in the description of the PR are bugs of the extraction plugin. Fixing them by changing CoqIDE is surprising. As Hugo noticed himself, this does not address the issue when using a different editor.

Please remove those legacy vernacular commands (Cd, Pwd). (After adding to the extraction plug-in a mechanism to conveniently chose where to write files.)

**Comment by herbelin** [comment]
> If I may share my thoughts, the issues that are referred to in the description of the PR are bugs of the extraction plugin. Fixing them by changing CoqIDE is surprising. As Hugo noticed himself, this does not address the issue when using a different editor.
> 
> Please remove those legacy vernacular commands (Cd, Pwd). (After adding to the extraction plug-in a mechanism to conveniently chose where to write files.)

Right, we could do something like that.

Do you have in mind a command of the form `Set Extraction Directory "foo".`? Do you have in mind that `Extraction "file" foo` fails if the directory is not set and `"foo"` is not absolute? Or that the same default as today is used (i.e. the directory where coqide is)? Or that the directory of the file being compiled is used, if any?

In any case, I believe this goes beyond the scope of this PR, so, probably, I'll let this PR stop to depend on #14720 in the meantime.

**Comment by vbgl** [comment]
Here is an experiment.

1. Create a fresh directory and launch coqide in said directory (many file managers have a feature “launch a command here”) and start hacking.
2. Write something like the following:
    > From Coq Require Extraction ZArith.
    > Separate Extraction BinInt.Z.testbit.
3. Don’t save, run all these commands.

Current behavior (released Coq): a few files are written in the fresh directory.

New behavior (after this PR): my home directory is littered with the extracted libraries.

To prevent this issue, there is a suggestion from Gaëtan above; here is an other one: don’t chdir when `file` is `None`.

**Comment by herbelin** [comment]
Hi, I agree that the PR as it is is not the right solution.

The situation seems more complex than initially thought as witnessed by issues such as #4953, #9148, #8649.

To concentrate on the exact initial problem, I propose to address only the MacOS issue for the released package.

By default, such package, when open with `open` (or a double-click) [sets](https://apple.stackexchange.com/questions/284754/what-is-the-default-working-directory-of-a-script-run-via-launchd) the cwd to the value `WorkingDirectory` in the `info.plist` of the application, and, if none, sets it to `/`. No variable such as `$HOME` can be given in the field.

I see at least two possible proposals with a preference for the 2nd:
- we set the WorkingDirectory to `/tmp`, but this is not so ideal since MacOS applications are supposed to have their own temporary files directory
- at run time, if on MacOS and the WorkingDirectory is `/`, we set it to `$HOME` (this is an overapproximation of what we really capture, since this would also be verified if I start coqide w/o `open` from the terminal, being in `/`). That seems reasonable though.

**Comment by herbelin** [comment]
What to think about this PR? I'am tempted to think that it is better to start the process in the home directory than in an unwritable directory. The situation is not so different after all from other MacOS applications which also have to choose a default directory.

**Comment by ppedrot** [comment]
I think that UIs should not mess with the current directory. It should only matter for open/save dialogues and even there, most interfaces I know of rely on higher-level mechanisms to store this information, so it's not even the right tool. I'd be inclined to close this PR.

**Comment by herbelin** [comment]
What would you propose as alternative?

**Comment by ppedrot** [comment]
For the UI, as per [the official documentation](https://people.gnome.org/~ryanl/gtk/html/GtkFileChooserDialog.html), literally nothing.
> Old versions of the file chooser's documentation suggested using gtk_file_chooser_set_current_folder() in various situations, with the intention of letting the application suggest a reasonable default folder. This is no longer considered to be a good policy, as now the file chooser is able to make good suggestions on its own. In general, you should only cause the file chooser to show a specific folder when it is appropriate to use gtk_file_chooser_set_filename(), i.e. when you are doing a File/Save As command and you already have a file saved somewhere.

For extraction, interactive use of Coq shouldn't start generating random files here and there. It should be a hard error except if some explicit option is set.

**Comment by herbelin** [comment]
> For extraction, interactive use of Coq shouldn't start generating random files here and there. It should be a hard error except if some explicit option is set.

Yes, I could do that in interactive use, `Extraction "foo"` fails if `foo` is a relative path.

Not sure that we need an option if it is going anyway to fail.

**Comment by Zimmi48** [comment]
Sorry, I haven't really followed this discussion but what about having `Extraction "foo"` open a dialog to ask the user where they want to save the file?

**Comment by Alizter** [comment]
What if we disallow this extraction command in CoqIDE all together? Just like we do for some options. That way we can have a dedicated menu item for extraction, which can in turn open a save dialog?

**Comment by herbelin** [comment]
These are good ideas but I'm afraid I have no time to explore myself how to trigger a dialog box from an arbitrary command nor to tell to a plugin how to inhibate a command for CoqIDE.

Also, the problem is only for the precompiled MacOS X package launched from its (unwritable) installation directory (/Applications/CoqIDE_8.14.0.app/Contents/MacOS/`.

**Comment by Zimmi48** [comment]
> What if we disallow this extraction command in CoqIDE all together? Just like we do for some options. That way we can have a dedicated menu item for extraction, which can in turn open a save dialog?

Disallowing or ignoring and displaying a warning? That's the behavior for the options you are talking about. The problem with disallowing entirely is that this command may make sense in batch mode and users may be bothered that they have to remove it to be able to execute their Coq file in CoqIDE.

**Comment by Zimmi48** [comment]
Turning this PR into a draft to reflect its status.

**Comment by ppedrot** [comment]
As per the author himself, this is not the right solution. No point in keeping a PR open which has no chance to make it, so closing it.

**Comment by herbelin** [comment]
The question of resolving the original issue is still pending and I somehow promised users of SF to solve it (even if temporary). There are also the general issues #4953, #9148 and #8649 about the commands writing to a file (did we get a conclusion about the long discussions there?).

Trying to summarize, what about the following proposal:
1. add a command line option setting an output dir (e.g. `-output-dir foo`) which needs to be set explicitly for `Extraction file`, `Redirect file` and `Print Universes file` to work
2. choose a semantics:
   - let all these commands fail when option is not set?
   - set a default for `coqc`? if yes, the same directory as where the `.v` is? (That would be consistent with writing the `.vo` in this directory)
   - set a default for `coqtop`, CoqIDE, VsCoq? If yes, maybe standard output together with an error message telling that writing a file needs the option to be set
3. really deprecate and remove `Cd`
4. reuse `Pwd` to output the value of `output-dir`??
5. forbid `Extraction`/`Redirect`/`Print Universes` to use absolute names, so as to answer #4953?

Note: slightly more complicated, we could also provide a way for plugins to declare command line options. Extraction could then declare its own directory? [I assume that we use `-require-import` first on the command line]

**Comment by Zimmi48** [comment]
This sounds like a good plan to me. For `coqc`, yes I'd say using the same directory as the `.v` file is a good solution (otherwise, it would create lots of compat issues). Whereas for CoqIDE et al. any solution is good (although I'd recommend a warning over an error to allow users to replay scripts without trouble if they don't care about the extracted files).

**Pull request closed without merging**